### ISSUE TITLE
Possible edge case bug in higher order function type inference

### ISSUE THREAD

**Issue by henriks** [Microsoft/TypeScript#30782]
**TypeScript Version:**  3.4.1 and 3.5.0-dev.20190405

**Search Terms:** higher order function type inference

**Code**

```ts
type F<T> = (t: T) => T;

function id<T>(f: F<T>): F<T> {
  return f;
}

function foo(s: string): string {
  return s;
}

function obj<T>(ks: { [K in keyof T]: F<T[K]> }): F<T> {
  return void 0 as any; // implementation is irrelevant
}

const o1 = obj({
  a: foo
});

const o2 = obj({
  a: id(foo) // compiler doesn't complain, vscode 1.33 does: Type 'F<string>' is not assignable to type 'F<{}>'.
});

const out1 = o2({ a: 123 }); // compiler doesn't catch this

const fields = {
  a: id(foo) // vscode is fine with this
};

const o3 = obj(fields);
const out2 = o3({ a: 123 }); // compiler catches this, as expected: Type 'number' is not assignable to type 'string'
```

**Expected behavior:**

I expected inference to work the same whether the `fields` object is defined inline or not.

**Actual behavior:**

Assigning the object to variable `fields` before calling `obj` with it makes inference work as expected, while defining the object inline in the call to `obj` breaks subsequent type checks.

**Playground Link:** [here](https://www.typescriptlang.org/play/#src=type%20F%3CT%3E%20%3D%20(t%3A%20T)%20%3D%3E%20T%3B%0D%0A%0D%0Afunction%20id%3CT%3E(f%3A%20F%3CT%3E)%3A%20F%3CT%3E%20%7B%0D%0A%20%20return%20f%3B%0D%0A%7D%0D%0A%0D%0Afunction%20foo(s%3A%20string)%3A%20string%20%7B%0D%0A%20%20return%20s%3B%0D%0A%7D%0D%0A%0D%0Afunction%20obj%3CT%3E(ks%3A%20%7B%20%5BK%20in%20keyof%20T%5D%3A%20F%3CT%5BK%5D%3E%20%7D)%3A%20F%3CT%3E%20%7B%0D%0A%20%20return%20void%200%20as%20any%3B%20%2F%2F%20implementation%20is%20irrelevant%0D%0A%7D%0D%0A%0D%0Aconst%20o1%20%3D%20obj(%7B%0D%0A%20%20a%3A%20foo%0D%0A%7D)%3B%0D%0A%0D%0Aconst%20o2%20%3D%20obj(%7B%0D%0A%20%20a%3A%20id(foo)%20%2F%2F%20compiler%20doesn't%20complain%2C%20vscode%201.33%20does%0D%0A%7D)%3B%0D%0A%0D%0Aconst%20out1%20%3D%20o2(%7B%20a%3A%20123%20%7D)%3B%20%2F%2F%20compiler%20doesn't%20catch%20this%0D%0A%0D%0Aconst%20fields%20%3D%20%7B%0D%0A%20%20a%3A%20id(foo)%0D%0A%7D%3B%0D%0A%0D%0Aconst%20o3%20%3D%20obj(fields)%3B%0D%0Aconst%20out2%20%3D%20o3(%7B%20a%3A%20123%20%7D)%3B%20%2F%2F%20compiler%20catches%20this%2C%20as%20expected) 

**Related Issues:** none

**Comment by RyanCavanaugh** [comment]
Moving type parameters to the return type only occurs when the return type of a function is syntactically a function type expression. A type alias like `F<T>` isn't one, so this doesn't occur.

**Comment by henriks** [comment]
But as far as I can tell, it does occur when I do the `const fields =` assign? The reason I raised an issue was the inconsistent behavior between the case where the object is assigned to a variable and then passed to the function, vs. when it's directly passed to the function.

**Comment by henriks** [comment]
This works as I would assume on 3.4.4, closing.
